libcnb/src/layer/handling.rs

```diff
@@ -0,0 +1,679 @@
+                    // the state on disk after we messed with it.
+                    read_layer(&context.layers_dir, layer_name.as_ref())?
+                        .ok_or(HandleLayerError::UnexpectedMissingLayer)
+                        .map_err(HandleLayerErrorOrBuildpackError::HandleLayerError)
+                }
+            }
+        }
+        Err(ReadLayerError::LayerContentMetadataParseError(_)) => {
+            match read_layer::<GenericMetadata, _, _>(&context.layers_dir, layer_name.as_ref()) {
+                Ok(Some(generic_layer_data)) => {
+                    let metadata_migration_strategy = layer
+                        .migrate_incompatible_metadata(
+                            context,
+                            &generic_layer_data.content_metadata.metadata,
+                        )
+                        .map_err(HandleLayerErrorOrBuildpackError::BuildpackError)?;
+
+                    match metadata_migration_strategy {
+                        MetadataMigration::RecreateLayer => {
+                            delete_layer(&context.layers_dir, layer_name.as_ref())?;
+                        }
+                        MetadataMigration::ReplaceMetadata(migrated_metadata) => {
+                            write_layer(
+                                &context.layers_dir,
+                                layer_name.as_ref(),
+                                generic_layer_data.env,
+                                LayerContentMetadata {
+                                    types: generic_layer_data.content_metadata.types,
+                                    metadata: migrated_metadata,
+                                },
+                            )?;
+                        }
+                    }
+
+                    handle_layer(context, layer_name, layer)
+                }
+                Ok(None) => Err(HandleLayerError::UnexpectedMissingLayer.into()),
+                Err(read_layer_error) => {
+                    Err(HandleLayerError::ReadLayerError(read_layer_error).into())
+                }
+            }
+        }
+        Err(read_layer_error) => Err(HandleLayerError::ReadLayerError(read_layer_error).into()),
+    }
+}
+
+fn handle_create_layer<B: Buildpack + ?Sized, L: Layer<Buildpack = B>>(
+    context: &BuildContext<B>,
+    layer_name: impl AsRef<str>,
+    layer: L,
+) -> Result<LayerData<L::Metadata>, HandleLayerErrorOrBuildpackError<B::Error>> {
+    let layer_dir = context.layers_dir.join(layer_name.as_ref());
+
+    let layer_result = layer
+        .create(context, &layer_dir)
+        .map_err(HandleLayerErrorOrBuildpackError::BuildpackError)?;
+
+    write_layer(
+        &context.layers_dir,
+        layer_name.as_ref(),
+        layer_result.env.unwrap_or_default(),
+        LayerContentMetadata {
+            types: layer.types(),
+            metadata: layer_result.metadata,
+        },
+    )?;
+
+    read_layer(&context.layers_dir, layer_name.as_ref())?
```

Comment:
What happens here with the `?` here ? I thought the `?` would return right away but then we're calling ok or and then mapping the error.

Reply:
Full code for context:

```rust
    read_layer(&context.layers_dir, layer_name.as_ref())?
        .ok_or(HandleLayerError::UnexpectedMissingLayer)
        .map_err(HandleLayerErrorOrBuildpackError::HandleLayerError)
```

`read_layer` returns a `Result<Option<..>>`. The `?` "resolves" the outer `Result`, returning the error if the `Result` is `Err`. The call to `ok_or` is on the inner `Option`. It converts the `Option` into _another_ `Result` that is either `Ok($VALUE_INSIDE_THE_OPTION>` or `Err(HandleLayerError::UnexpectedMissingLayer)`. The final `map_error` wraps the error in `HandleLayerErrorOrBuildpackError::HandleLayerError`.

We do this to "promote" the inner `Option` into a specific error should it be `None`.

Reply:
Cool, thanks! It's hard to read some of this code in PR mode since similar methods (map/map_err) exist on different types, this might have been more clear with an IDE that is also annotating types